prboom2/src/gl_shader.c

```diff
@@ -53,217 +54,534 @@
+    return;
+  }
 
-  idx = GLEXT_glGetUniformLocationARB(sh_indexed->hShader, "tex");
-  GLEXT_glUniform1iARB(idx, 0);
+  if (src->size == src->cap)
+  {
+    if (src->cap == 0)
+      src->cap = 1024;
+    else
+      src->cap *= 2;
 
-  idx = GLEXT_glGetUniformLocationARB(sh_indexed->hShader, "colormap");
-  GLEXT_glUniform1iARB(idx, 2);
+    src->strs = Z_Realloc(src->strs, src->cap * sizeof(*src->strs));
+    src->lens = Z_Realloc(src->lens, src->cap * sizeof(*src->lens));
+  }
 
-  GLEXT_glUseProgramObjectARB(0);
+  src->strs[src->size] = str;
+  src->lens[src->size] = len;
+  ++src->size;
 }
 
-static void get_fuzz_shader_bindings()
+static void ShaderLookup(const char* name, GLchar const** text, GLint* len)
 {
-  int idx;
+  int lump = W_CheckNumForName2(name, ns_prboom);
 
-  fuzz_unifs.tex_d_index = GLEXT_glGetUniformLocationARB(sh_fuzz->hShader, "tex_d");
-  fuzz_unifs.ratio_index = GLEXT_glGetUniformLocationARB(sh_fuzz->hShader, "ratio");
-  fuzz_unifs.seed_index = GLEXT_glGetUniformLocationARB(sh_fuzz->hShader, "seed");
+  if (lump == LUMP_NOT_FOUND)
+    I_Error("Could not find shader source: %s\n", name);
 
-  GLEXT_glUseProgramObjectARB(sh_fuzz->hShader);
+  *text = W_LumpByNum(lump);
+  *len = W_LumpLength(lump);
+}
 
-  idx = GLEXT_glGetUniformLocationARB(sh_fuzz->hShader, "tex");
-  GLEXT_glUniform1iARB(idx, 0);
+static dboolean GetLine(GLchar const** text, GLint* len, GLchar const** line, GLint* llen)
+{
+  *line = *text;
+  *llen = 0;
+  while (*len)
+  {
+    --*len;
+    ++*llen;
+    if (*(*text)++ == '\n')
+      return true;
+  }
+  return *llen != 0;
+}
 
-  idx = GLEXT_glGetUniformLocationARB(sh_fuzz->hShader, "fuzz");
-  GLEXT_glUniform1iARB(idx, 1);
+static dboolean FindCharAfter(const GLchar* str, GLint len, char c, GLchar const** found, GLint* flen)
+{
+  while (--len)
+  {
+    if (*str == c)
+    {
+      *found = str + 1;
+      *flen = len;
+      return true;
+    }
+    ++str;
+  }
 
-  GLEXT_glUseProgramObjectARB(0);
+  return false;
 }
 
-void glsl_Init(void)
+static void SkipWhitespace(GLchar const** str, GLint* len)
 {
-  sh_indexed = gld_LoadShader("glvp", "glfp_idx");
-  get_indexed_shader_bindings();
+  while (*len && isspace(**str))
+  {
+    --*len;
+    ++*str;
+  }
+}
 
-  sh_fuzz = gld_LoadShader("glvp", "glfp_fuzz");
-  get_fuzz_shader_bindings();
+#define CSLEN(x) (sizeof((x)) - 1)
+
+static void ShaderSrcAppendDefine(shdr_src_t* src, const shdr_def_t* def)
+{
+  static const char cdefine[] = "#define ";
+  static const char cspace[] = " ";
+  static const char cnl[] = "\n";
+
+  ShaderSrcAppend(src, cdefine, CSLEN(cdefine));
+  ShaderSrcAppend(src, def->name, strlen(def->name));
+  ShaderSrcAppend(src, cspace, CSLEN(cspace));
+  ShaderSrcAppend(src, def->value, strlen(def->value));
+  ShaderSrcAppend(src, cnl, CSLEN(cnl));
 }
 
-static int ReadLump(const char *filename, const char *lumpname, char **buffer)
+static void ShaderSrcProcess(shdr_src_t* src, const GLchar* text, GLint len,

```

Comment:
Basically, we have our own preprocessor?

Reply:
For `#include`, yeah.  Everything else is handled by OpenGL.